File: data_structures/trees/binary_trees/binary_search_trees/binary_search_tree.js
class Node {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}


class BinarySearchTree {
  constructor() {
    this.root = null;
  }

  // O(log N)
  insert(value) {
    let newNode = new Node(value);
    if (this.root === null) {
      this.root = newNode;
      return this;
    }
    let currentNode = this.root;
    while (true) {
      if (value < currentNode.value) {
        if (currentNode.left === null) {
          currentNode.left = newNode;
          return this;
        }
        currentNode = currentNode.left;
      }
      else if (value > currentNode.value) {
        if (currentNode.right === null) {
          currentNode.right = newNode;
          return this;
        }
        currentNode = currentNode.right;
      }
      else {
        return this;
      }
    }
  }

  // O(log N)
  search(value) {
    let currentNode = this.root;
    while (currentNode !== null) {
      if (value === currentNode.value) {
        return currentNode;
      }
      if (value < currentNode.value) {
        currentNode = currentNode.left;
      }
      else {
        currentNode = currentNode.right;
      }
    }
    return null;
  }

  // O(log N)
  remove(value) {
    let currentNode = this.root;
    let parentNode = null;

    while (currentNode !== null) {
      if (value < currentNode.value) {
        parentNode = currentNode;
        currentNode = currentNode.left;
      }
      else if (value > currentNode.value) {
        parentNode = currentNode;
        currentNode = currentNode.right;
      }
      else if (currentNode.value === value) {

        // Option 1: No right child
        if (currentNode.right === null) {
          if (parentNode === null) {
            this.root = currentNode.left;
          }
          else {
            // if current value < parent, make current node left child a child of parent
            if (currentNode.value < parentNode.value) {
              parentNode.left = currentNode.left;

            }
            // if current value > parent, make left child a right child of parent
            else if (currentNode.value > parentNode.value) {
              parentNode.right = currentNode.left;
            }
          }
        }

        //Option 2: Right child which doesnt have a left child
        else if (currentNode.right.left === null) {
          currentNode.right.left = currentNode.left;
          if (parentNode === null) {
            this.root = currentNode.right;
          }
          else {
            // if parent > current, make right child of the left the parent
            if (currentNode.value < parentNode.value) {
              parentNode.left = currentNode.right;

            }
            // if parent < current, make right child a right child of the parent
            else if (currentNode.value > parentNode.value) {
              parentNode.right = currentNode.right;
            }
          }
        }

        //Option 3: Right child that has a left child
        else {

          //find the Right child's left most child
          let leftmost = currentNode.right.left;
          let leftmostParent = currentNode.right;
          while (leftmost.left !== null) {
            leftmostParent = leftmost;
            leftmost = leftmost.left;
          }

          //Parent's left subtree is now leftmost's right subtree
          leftmostParent.left = leftmost.right;
          leftmost.left = currentNode.left;
          leftmost.right = currentNode.right;

          if (parentNode === null) {
            this.root = leftmost;
          }
          else {
            if (currentNode.value < parentNode.value) {
              parentNode.left = leftmost;
            }
            else if (currentNode.value > parentNode.value) {
              parentNode.right = leftmost;
            }
          }
        }
        return true;
      }
    }
    return false;
  }

}

File: data_structures/trees/binary_trees/binary_search_trees/binary_search_tree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


class BinarySearchTree:
    def __init__(self):
        self.root = None
        self.length = 0

    # O(log N)
    def insert(self, value):
        new_node = Node(value)
        if self.root is None:
            self.root = new_node
            return self
        current_node = self.root
        while True:
            if value < current_node.value:
                if current_node.left is None:
                    current_node.left = new_node
                    return self
                current_node = current_node.left
            elif value > current_node.value:
                if current_node.right is None:
                    current_node.right = new_node
                    return self
                current_node = current_node.right
            else:
                return self

    # O(log N)
    def search(self, value):
        current_node = self.root
        while current_node is not None:
            if value == current_node.value:
                return current_node
            elif value < current_node.value:
                current_node = current_node.left
            else:
                current_node = current_node.right

    def remove(self, value):
        current_node = self.root
        parent_node = None
        while current_node is not None:
            if value < current_node.value:
                parent_node = current_node
                current_node = current_node.left
            elif value > current_node.value:
                parent_node = current_node
                current_node = current_node.right
            else:
                # Case 1: When deleting the root node
                if parent_node is None:
                    self.root = None
                
                # Case 2: When the node being deleted has no child, just simply delete it
                elif current_node.left is None and current_node.right is None:
                    if current_node.value < parent_node.value:
                        parent_node.left = None
                    else:
                        parent_node.right = None
                
                # Case 3: If it has only a left child, plug that child in the spot where the node being deleted was
                elif current_node.left is not None and current_node.right is None:
                    if current_node.value < parent_node.value:
                        parent_node.left = current_node.left
                    else:
                        parent_node.right = current_node.left
                
                # Case 4: Or if it has only a right child, plug that child in the spot where the node being deleted was  
                elif current_node.right is not None and current_node.left is None:
                    if current_node.value < parent_node.value:
                        parent_node.left = current_node.right
                    else:
                        parent_node.right = current_node.right
                
                # Case 5: When deleting a node with two children, replace the deleted node
                # with the successor node. The successor node is the child node whose 
                # value is the least of all values that are greater than the deleted node.
                else:
                    # To find the successor node isit the right child of the to-be-deleted value
                    # and then keep on visiting the left child of each subsequent child until
                    # there are no more left children. The bottom value is the successor node.
                    successor_node = current_node.right
                    parent_of_successor = current_node
                    while successor_node.left is not None:
                        parent_of_successor = successor_node
                        successor_node = successor_node.left

                    # If the successor node has a right child, take that child and turn
                    # it into the left child of the former parent of the successor node.
                    if successor_node.right is not None:
                        parent_of_successor.left = successor_node.right
                    
                    if current_node.value < parent_node.value:
                        parent_node.left = successor_node
                    else:
                        parent_node.right = successor_node
                    successor_node.left = current_node.left
                    successor_node.right = current_node.right
            return True
        return False